### ISSUE TITLE
Model Parallelism Memory usage issue

### ISSUE THREAD

**Issue by reger-men** [tensorflow/tensorflow#21522]
------------------------

### System information
- **Have I written custom code**: Yes
- **OS Platform and Distribution**: Linux Ubuntu 16.04
- **Mobile device**: N/A
- **TensorFlow installed from**: From binary
- **TensorFlow version**: 1.9
- **Python version**: 3.6.1
- **Bazel version**: N/A
- **GCC/Compiler version (if compiling from source)**: 5.4.0
- **CUDA/cuDNN version**: 9.0.176
- **GPU model and memory**: Tested on 4X GeForce GTX 1080 with 10GB
- **Exact command to reproduce**: `CUDA_VISIBLE_DEVICES=0,1,2,3 python3 train.py --data-dir=Datasets_Path/cityscapes/ --input-size="1000,1000" --batch-size=1`

### Describe the problem
I run the above Model training on single GPU and on multiple GPUs and it looks like the memory is being allocated several times. The model size is about 10,7 GB. By distributing over several GPUs the total model size should not be increased. But as you can see it almost triples. Thereby I can not profit from model parallelism. Even a input size of 1200 x 1200 return an OOM error.

The following screenshots show the memory usage on different GPUs:

### On GPU:0
![gpu0](https://user-images.githubusercontent.com/8779942/43915687-c2372af4-9c0b-11e8-8eac-c7d3d28b7844.png)

### On GPU:0 and 1
![gpu01](https://user-images.githubusercontent.com/8779942/43915680-bda22de0-9c0b-11e8-957e-79321a8a834b.png)

### On GPU:0, 1 and 3
![gpu012](https://user-images.githubusercontent.com/8779942/43915644-a35d67ec-9c0b-11e8-8cd6-e79e948699c4.png)

### On GPU:0, 1,2 and 3
![gpu0123](https://user-images.githubusercontent.com/8779942/43915637-9b08e7a6-9c0b-11e8-9621-5e679af4bae6.png)

### Source code / logs
[Source code](https://github.com/reger-men/PSPNet-Tensorflow-ModelParallelism)

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
Bazel version
Mobile device

**Comment by aselle** [comment]
@zheng-xq , could you ptal.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by reger-men** [comment]
It's an issue and I'm still waiting for a response.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
@reger-men : Sorry, this seemed to have fallen through the triage cracks.

I took a quick look at the attached code, but there is a lot of it :). Is it possible to write a more succinct example demonstrating the problem? In general though, this could happen if there are model parameters that are being read across GPUs. I'd have to spend a lot more time looking at the framework you're using to be sure that this isn't the case, but if you could write a shorter snippet, that would be helpful. Thanks.

(@azaks2 FYI)

**Comment by reger-men** [comment]
@asimshankar Thank you for taking the time for this issue. 
I just created a new branch **cifar10** its a short example that shows a similar problem on model Parallelism.
You can change the script argument as you need _(eg. input size, batch size, etc)_

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by msabony1966** [comment]
guys,
This is serious issue. really appreciate if you could fix it
Thanks for your support
Best Regards
Mazda

**Comment by asimshankar** [comment]
@reger-men : Is it possible to get a shorter snippet that demonstrates the problem (ideally one that doesn't require downloading the dataset or figuring out how exactly to run the command in your repository). Could you elaborate on what you mean by "memory is being allocated multiple times"? @msabony1966 : Perhaps you have another, simpler, example demonstrating the problem?

Note that TensorFlow doesn't return GPU memory to the system when it allocates it. So the memory usage you see in `nvidia-smi` will be the peak memory utilized on each GPU, which in-turn would need to account for both model parameters hosted on the device as well as the size of intermediate tensors computed along the way.

Is it possible that even with model parallelism you're doing some operations that require parameters hosted on one GPU to be transferred to another GPU or something? Or that there are intermediate tensors computed on a single GPU that are larger than the memory footprint of that single GPU?

**Comment by reger-men** [comment]
@asimshankar as you can see in the scrreshots, when I run the training on only 1 GPU the allocated memory is about 10GB. Run the same model with the same parameters (batch_size, input_size, etc.) on multiple GPU required more than 24GB. How to explain that?

My understanding the using of model parallelism, if my model is very large and does not fit into a single GPU. Therefore I distribute the model on different GPUs to make training possible. Of course, the data here has to be exchanged between the GPUs for both forward and backward propagation and the only way is through the host. But this does not explain why too much memory is allocated on the GPUs.

I already insert a new branch in the repo ["cifar10"](https://github.com/reger-men/PSPNet-Tensorflow-ModelParallelism.git). You can also observe the same problem there.

**Comment by asimshankar** [comment]
@reger-men : The memory consumed by a model depends on the computation in the model, not just variable placement. For example, consider this trivial computation:

```python
x = tf.random_uniform([1000, 1000, 1000]) # A ~4GB tensor
y = tf.add(x, 1.)
z = tf.add(y, 1.)

with tf.Session() as sess:
  sess.run(z)
```

This computation will consume 4GB. If I split it across 2 GPUs with something like:

```python
with tf.device("/gpu:0"):
  x = tf.random_uniform([1000, 1000, 1000])
  y = tf.add(x, 1.)
with tf.device("/gpu:1"):
  z = tf.add(y, 1.)

with tf.Session() as sess:
  sess.run(z)
```

it will consume 4GB on each GPU since a 4GB tensor needs to be materialized in the memory of each device.

When distributing the computation across devices the memory usage characteristics would depend on how the computation (and thus intermediate tensor values) is distributed. Consider another trivial program:

```python
a = tf.random_uniform([1000, 1000, 1000]) # ~4GB
b = tf.random_uniform([1000, 1000, 1000]) # ~4GB
c = tf.random_uniform([1000, 1000, 1000]) # ~4GB
d = tf.add_n([a, b, c])
e = tf.reduce_sum(d)

with tf.Session() as sess:
  sess.run(e)
```

will consume ~12GB on one GPU (as 3 4GB tensors are materialized), but if you split it across 3 GPUs like:

```python
with tf.device("/gpu:0"):
  a = tf.random_uniform([1000, 1000, 1000])
with tf.device("/gpu:1"):
  b = tf.random_uniform([1000, 1000, 1000])
with tf.device("/gpu:2"):
  c = tf.random_uniform([1000, 1000, 1000])
  d = tf.add_n([a, b, c])
  e = tf.reduce_sum(d)

with tf.Session()
  sess.run(e)
```

then it will consume ~4GB on GPU 0, ~4GB on GPU 1 and ~12GB on GPU 2 (since a 4GB tensor is materialized on each GPU, but additionally a 4GB tensor is shipped from GPU 0 to GPU 2 and another 4GB tensor is shipped from GPU 1 to GPU 2).

But if you split it like:

```python
with tf.device("/gpu:0"):
  a = tf.random_uniform([1000, 1000, 1000])
with tf.device("/gpu:1"):
  b = tf.random_uniform([1000, 1000, 1000])
  b_1 = tf.add(a, b)
with tf.device("/gpu:2"):
  c = tf.random_uniform([1000, 1000, 1000])
  d = tf.add(b_1, c)
  e = tf.reduce_sum(d)

with tf.Session()
  sess.run(e)
```

it will consume ~4GB on GPU 1, ~8GB on GPU 1 and ~8GB on GPU 2, and if you split it like:

```python
with tf.device("/gpu:0"):
  a = tf.random_uniform([1000, 1000, 1000])
  a_1 = tf.reduce_sum(a)
with tf.device("/gpu:1"):
  b = tf.random_uniform([1000, 1000, 1000])
  b_1 = tf.reduce_sum(b)
with tf.device("/gpu:2"):
  c = tf.random_uniform([1000, 1000, 1000])
  c_1 = tf.reduce_sum(c)
  e = tf.add_n([a_1, b_1, c_1])

with tf.Session()
  sess.run(e)
```

it will consume ~4GB on each device.

Does that make sense?

Note that the TensorFlow runtime is free to optimize the graph execution and in theory could do such memory saving transformations itself. However, as of _today_, the examples I've described above should have the characteristics described.

I haven't had the time to dig into your sample code to understand the custom framework being used to build the network and even so running it seems to require setting up a dataset etc. first. You might get some hint of where the memory is coming from by looking at the shapes of tensors being shipped across devices.  From a cursory look at https://github.com/reger-men/PSPNet-Tensorflow-ModelParallelism/blob/279285d4e9ea9741d7208a0e4c37d3ecf42d55c7/model.py#L75 it seems that the tensors `conv3_1_1x1_proj_bn1` and `conv3_1_1x1_increase_bn` are being shipped from GPU 0 to 1, so I'd look at the shapes of those tensors, and any others going across devices. Besides the tensors being shipped, there is also the question of what the shapes and sizes of intermediate tensors being computed on the device is. 

Again, the code you have is a lot to go through to figure this out. If you can reduce it to a [minimal, complete, verifiable example](https://stackoverflow.com/help/mcve) that will make it easier to figure out if there is a real problem or if the behavior you're seeing is justified given how the model is being distributed across devices.

Hope that helps.

**Comment by reger-men** [comment]
@asimshankar Thank you for the clarification.
I have considered what you have described and yet the memory allocation was well above the theoretical value. 

Let us look at the following example (Btw, I have already uploaded it under Cifar10 Branch): 
At https://github.com/reger-men/PSPNet-Tensorflow-ModelParallelism/blob/cifar10/include/model.py#L26

With consideration of the input (`10x32x3x32x32`) in this order(batch_size, preview filter number, channel, height, width)  which gives about 4MB including the processing (`conv 3x3 64 filters, Max_pool 2x2 and dropout of 0.25`), we come to about **40MB**. The memory allocation on GPU1, where this part run is about **2GB**. The model just contains 5 layers.

Thanks,

**Comment by gadagashwini** [comment]
Hi @reger-men,
Tensorflow version 2.x support `tf.distribute.Strategy`. 
[tf.distribute.Strategy](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy) is a TensorFlow API to distribute training across multiple GPUs, multiple machines. Using this API, you can distribute your existing models and training code with minimal code changes. For more information refer [Distributed training with TensorFlow](https://www.tensorflow.org/guide/distributed_training). Thank you!
